### PR TITLE
Implement Poseidon (Second take)

### PR DESCRIPTION
**Pull request by ChihChengLiang**
We do the Poseidon 11 using the way described in #88

### CODE REVIEW

File: circuits/circom/batchUpdateStateTree.circom [hunk 1]
```diff
@@ -54,7 +54,7 @@ template BatchUpdateStateTree(
         // if msg_tree_batch_end_index < message_indices[i], use msg_tree_batch_end_index
         // if msg_tree_batch_end_index >= message_indices[i], use message_indices[i]
 
-        message_indices[i] = msg_tree_batch_start_index + i;
+        message_indices[i] <== msg_tree_batch_start_index + i;
```
Comment (ChihChengLiang) [inline circuits/circom/batchUpdateStateTree.circom:57]: Hummm, the reason for this diff is zkutils complaining this
```
ERROR at /Users/liangcc/projects/EAPL/semaphore_projects/maci/circuits/circom/batchUpdateStateTree.circom:57,8-57,59   Cannot assign to a signal with `=` use <-- or <== ops
```

Comment (weijiekoh) [reply circuits/circom/batchUpdateStateTree.circom:57]: Could it be a different version of `circom`?

```bash
~/maci/circuits$ rm build/batchUstCircuit.json && ./scripts/buildBatchUpdateStateTreeSnark.sh
Compiling circom/prod/batchUpdateStateTreeVerifier.circom...
Generated build/batchUstCircuit.json
Params file exists. Skipping setup.
Generating Solidity verifier...
Copying BatchUpdateStateTreeVerifier.sol to contracts/sol.
~/maci/circuits$ npx circom --version
0.0.34
```

Comment (ChihChengLiang) [reply circuits/circom/batchUpdateStateTree.circom:57]: Thanks, let me check

Comment (ChihChengLiang) [reply circuits/circom/batchUpdateStateTree.circom:57]: humm, I'm using the same version too.
```
$  npx circom --version
0.0.34
```

File: contracts/ts/__tests__/Hasher.test.ts [hunk 2]
```diff
@@ -27,38 +29,46 @@ describe('Hasher', () => {
...
+        })
     })
 
     it('maci-crypto.hashLeftRight should match hasher.hashLeftRight', async () => {
         const left = genRandomSalt()
         const right = genRandomSalt()
         const hashed = hashLeftRight(left, right)
 
-        const onChainHash = await hasherContract.hashLeftRight(left.toString(), right.toString())
+        const onChainHash = await t3Contract.poseidon([left.toString(), right.toString()])
```
Comment (ChihChengLiang) [inline contracts/ts/__tests__/Hasher.test.ts:49]: Hummm, strange. `poseidon.createHash(3, 8, 49, 'poseidon')` disagree with `poseidon.createCode(3, 8, 49, 'poseidon')`

**Comment by weijiekoh** [comment]
Feel free to delete the MiMC circuit tests :D

File: circuits/circom/verify_signature.circom [hunk 3]
```diff
@@ -45,7 +45,7 @@ template EdDSAMiMCSpongeVerifier_patched() {
   hash.k <== 0;
```
Comment (ChihChengLiang) [inline circuits/circom/verify_signature.circom:45]: I still keep the MiMC sponge in the signature verification here. Let's change it when we have more knowledge on #98

File: circuits/circom/verify_signature.circom [hunk 4]
```diff
@@ -118,11 +118,13 @@ template VerifySignature(k) {
 
   signal output valid;
   
-  component M = Hasher(k);
-  M.key <== 0;
+  component M = Hasher11();
```
Comment (ChihChengLiang) [inline circuits/circom/verify_signature.circom:121]: We are using Hasher11 here because it just hashes a message into a snarkBigInt. Nothing non-trivial to do with signature.

File: crypto/ts/index.ts [hunk 5]
```diff
@@ -68,22 +77,50 @@ const hash = (plaintext: Plaintext): SnarkBigInt => {
     return mimcsponge.multiHash(plaintext, 0, 1)
```
Comment (ChihChengLiang) [inline crypto/ts/index.ts:77]: I'm keeping this `hash` because it is used in `sign`

File: circuits/circom/verify_signature.circom [hunk 6]
```diff
@@ -118,11 +118,13 @@ template VerifySignature(k) {
...
   for (var i = 0; i < k; i++){
     M.in[i] <== preimage[i];
   }
+  for (var i = k; i < 11; i++){
```
Comment (weijiekoh) [inline circuits/circom/verify_signature.circom:125]: I think we should do another PR to set a constant number of inputs for `VerifySignature`: see #99

**Comment by weijiekoh** [comment]
With MiMC:
```
$ npx snarkjs info -c build/batchUstCircuit.json 
# Wires: 264161
# Constraints: 264160
# Private Inputs: 159
# Public Inputs: 19
# Outputs: 1

$ npx snarkjs info -c build/qvtCircuit.json 
# Wires: 141366
# Constraints: 141266
# Private Inputs: 104
# Public Inputs: 4
# Outputs: 1
```

Gas cost for deploying `MACI.sol` with tree depths 4, 4, 4: 6147852

With Poseidon:

```
$ npx snarkjs info -c build/batchUstCircuit.json 
# Wires: 90229
# Constraints: 90232
# Private Inputs: 159
# Public Inputs: 19
# Outputs: 1

$ npx snarkjs info -c build/qvtCircuit.json 
# Wires: 22242
# Constraints: 22142
# Private Inputs: 104
# Public Inputs: 4
# Outputs: 1
```

Gas cost for deploying `MACI.sol` with tree depths 4, 4, 4: 6836062

File: circuits/README.md [hunk 7]
```diff
@@ -5,13 +5,32 @@
...
-| t = 3      |                  2 | Build Merkle tree with HashLeftRight   |
-| t = 6      |                  5 | Hash stateLeaf                         |
+| Poseidon t | number of elements | Use cases                            |
+| ---------- | -----------------: | ------------------------------------ |
+| t = 3      |                  2 | Build Merkle tree with HashLeftRight |
+| t = 6      |                  5 | Hash stateLeaf                       |
 
 The Message has 11 elements and we hash it with a Hasher11, which is a combination of t3 and t6.
+
+## Why `t := n+1`?
```
Comment (weijiekoh) [inline circuits/README.md:19]: Awesome :D :D
